03/haenniro.ml

```diff
@@ -0,0 +1,33 @@
+open Assert
+
+let prefix_fn fn = "./compiler-design-eth-tests/03/haenniro/" ^ fn
+
+let exec_e2e_ast ll_ast extra_files =
+  let output_path = !Platform.output_path in
+  let dot_s_file = Platform.gen_name output_path "test" ".s" in
+  let exec_file = Platform.gen_name output_path "exec" "" in
+  let asm_ast = Backend.compile_prog ll_ast in
+  let asm_str = X86.string_of_prog asm_ast in
+  let _ = Driver.write_file dot_s_file asm_str in
+  let _ = Platform.link (dot_s_file::extra_files) exec_file in
+  let result = Driver.run_executable "" exec_file in
+  let _ = Platform.sh (Printf.sprintf "rm -f %s %s" dot_s_file exec_file) Platform.ignore_error in
+  let _ = Platform.verb @@ Printf.sprintf "** Executable exited with: %d\n" result in
+  Int64.of_int result
+
+
+let exec_e2e_file path extra_files =
+  let ast = Driver.parse_ll_file path in
+  exec_e2e_ast ast extra_files
+
+let executed tests =
+  List.map (fun (fn, extra_files, ans) ->
+      fn, assert_eqf (fun () -> exec_e2e_file (prefix_fn fn) (List.map prefix_fn extra_files)) ans)
+    tests
```

Comment:
I think you can use `executed_c_link` for this from `Gradedtests` module

Reply:
absolutely! completely missed that function haha